Glossaries change layout

t-cordonnier opened the pull request:
Enable to:
- set colors associated to source, translation and comment of glossaries
- choose layout between actually 4 layouts
- define new layout classes as plugins, which can be loaded from external JAR

amake commented on src/org/omegat/gui/glossary/GlossaryTextArea.java:
```diff
@@ -167,10 +167,30 @@ public Component getComponentToOverlay() {
+                    entryRenderer = r;
+                }
+            }
+        } catch (Exception e) {
+        }
```
What is the exception that could be thrown here?

t-cordonnier replied:
You are true, I just removed it.

amake commented on src/org/omegat/gui/glossary/TransTipsMarker.java:
```diff
@@ -68,8 +67,9 @@
 
         List<Mark> marks = new ArrayList<Mark>();
 
+        IGlossaryRenderer entryRenderer = Core.getGlossary().getLocalRenderer();
```
Is there really a need to fetch the renderer out of the Glossary Pane? It seems like unnecessary coupling.

Is there a reason not to simply have a separate instance here?

t-cordonnier replied:
OK, for me the important is to share the algorithm giving renderer by name (because else it could happen that you accidentally do a modification in GlossaryTextArea and forget to replicate it to TransTips). But if we consider that I accept the idea of creating the class GlossaryRenderers (question 1), then I think that the common method giving the selected renderer should be there.

amake replied:
I've moved the preferred renderer into `GlossaryRenderers`.

amake reviewed (commented):
Sorry for the very, very long delay on this.

If you're still willing to work on this with me, I have some questions.

t-cordonnier commented on src/org/omegat/core/Core.java:
```diff
@@ -131,6 +132,7 @@ private Core() {
             .synchronizedList(new ArrayList<String>());
 
     private static final List<IMarker> MARKERS = new ArrayList<IMarker>();
+    private static final List<IGlossaryRenderer> GLOS_RENDERERS = new ArrayList<IGlossaryRenderer>();
```
No problem for me. The only pattern I tried to follow was consistency with existing code.
I understand that you want to change some patterns in the code of OmegaT, but I think it would be better if you could document it somewhere, instead of saying it individually to developers once they wrote code following old rules. IssueProviders.java is a good reference but its existence is not enough to say that you want to work like that in the future: you must write it somewhere.

Globally I agree with your thoughts about coupling in the message you quote. My reaction in june was not about reducing dependency to Core, but about adding useless dependency between glossaries and all MT engines, while only one really needed it; and after I saw your explanation, once I had all the constraints in mind, I proposed a solution.
And I also emphased in the next message that there are plenty of old code which does not respect the new rules: not only you should document your new wishes, but you could almost change some of old classes, so that you can in your documentation put an example about how we did before and how we do now.

When you say "remove existing stuff", do you mean that you would like to change the pattern for other plugin types as well, and take the risk to break compatibility with external plugins? If so, I would advice that you open a discussion in the developers list. I have some ideas of improvements about the plugins system, but I would better share them in the users list, and preferably in an active discussion, rather than starting it myself and receive no answer because everybody is busy about something else. 

That said, yes, I agree with the principle of this change, I would only have preferred to know it before.

PS. Other questions I will answer later. I wrote this code one year ago, I really need some time to remember what I had in mind at this time.

t-cordonnier commented on src/org/omegat/gui/glossary/IGlossaryRenderer.java:
```diff
@@ -25,14 +25,133 @@
+
+    // --------------- Rendering methods -----------------
+
+    boolean wantsMerge();
```
Sorry, I had forgotten to commit some piece of code. Have a look now at GlossarySearcher.java, line 285.

amake replied:
Thanks. I understand now.

But merging/no-merging seems like a concern that is not related to the glossary renderer, since the only place it takes effect is *outside* of the rendering.

On the other hand I don't like adding a million toggles for every little thing. In fact I wonder if I said in a previous discussion that I would *not* want merging/no-merging to be a separate control. If I said that, it was probably under the (mistaken) assumption that merging takes place at render time rather than before it.

Seeing the result, however, I think adding "merge" and "no-merge" variants for every renderer is strictly worse than just having a separate toggle.

amake replied:
I've moved this flag to a separate preference.

t-cordonnier replied:
I totally agree with this... but precisely, the problem is that _you_ said explicitly the contrary one year ago:
https://sourceforge.net/p/omegat/mailman/message/36777299/

In one of your messages you said that this feature was "orthogonal" to glossary layout: now I think that you understand that when two features are orthogonal, joining them in one means that the number of values in drop box is the scalar product of number of possibilities for both options!

The only reason why I had implemented like this is precisely the message I am quoting right now. But at the same time I had implemented in DGT a version with checkbox, exactly as what you are doing now. I always work like that, I always implement first in DGT and I advice to people, including you, to test the features in DGT - which does not mean that you have to use it after that. If you had tested the feature, instead of only speaking theoritically, you would have understood immediately that the checkbox did something totally different than glossary layout!

amake commented:
Thanks for updating the PR and for your answers.

I'm worried that if we have too much back-and-forth then I'll run out of time and this will drag on even longer. So I'm going to make the changes that I'd like to see and then push them for discussion. I'll let you know when I'm done.

amake commented:
I've made my changes.

1. I moved storage of `IGlossaryRenderer`s to the `GlossaryRenderers` class
2. I changed the built-in renderers to no longer be plugins.

    We really need to be able to assume that `DefaultGlossaryRenderer` exists. `DictionaryGlossaryRenderer` could be a plugin, but I don't see a need for it to be one right now.
3. I made merging/not-merging a separate option
4. I reverted the default glossary colors because I found them too garish.

    I would be OK with changing the defaults to something not too far from the current scheme, for instance making the "note" a shade of gray to de-emphasize it a bit. But I feel pretty strongly that bright colors need to be opted into by the user. I also suspect that there is not going to be a consensus on what colors should be used. That means that users who want color currently need to manually customize the colors, and I'm OK with that.

Please let me know what you think.

t-cordonnier commented:
>     1. I moved storage of `IGlossaryRenderer`s to the `GlossaryRenderers` class

OK for me.
I did a last change in this class: since GlossaryTextArea and TransTipsMarker do very frequent calls to the method getPreferredGlossaryRenderer(), better do not browse the list each time.

> 
>     2. I changed the built-in renderers to no longer be plugins.
>        We really need to be able to assume that `DefaultGlossaryRenderer` exists. `DictionaryGlossaryRenderer` could be a plugin, but I don't see a need for it to be one right now.
> 

OK for me

>     3. I made merging/not-merging a separate option
> 

Yes. See my comment about this point in a separate message.

>     4. I reverted the default glossary colors because I found them too garish.
>        I would be OK with changing the defaults to something not too far from the current scheme, for instance making the "note" a shade of gray to de-emphasize it a bit. But I feel pretty strongly that bright colors need to be opted into by the user. I also suspect that there is not going to be a consensus on what colors should be used. That means that users who want color currently need to manually customize the colors, and I'm OK with that.
> 

That is your choice, since it is now configurable it can be changed later, so no problem for me.
I don't know how I could have choosen less "garish" colors, except, as you say, with some shades of gray. Here is how I did the choice:
a) Be consistent with other colors used in the application. My idea was to use the same colors as the ones used in the editor (blue for source, green for translation, purple for notes). But now I realize that DGT did not use the same color scheme as standard OmegaT and I forgot to replicate the change.
b) Since in the editor we use background colors while the glossary pane uses font colors over white background, I must use darker colors (reason why I use "dark green", because standard green seems difficult to read on white screen)
c) Use standard colors, such as Color.blue, instead of a self-made combination, so that they don't immediately appear as strange for the users

What you say about consensus is not necessarily false, but it did not prevent from having lot of colors in the past in the editor, so I think a discussion should be opened in the users list. That said, we can commit like that and change later if a consensus arrives.

amake commented:
Thanks for your help and your patience with this PR. I've squashed and pushed to `master` on SourceForge.

I'll be happy to revisit details based on feedback.
